### PR TITLE
Add opm-build-push target

### PR DESCRIPTION
**Pull request by rporres**
* Opt-in target, we are not ready to force anyone to use this
* Adds a script to build catalog using opm
* Adds authentication to docker-push (breaking change)
* Depends on a local `hack/generate-bundle-contents.py` script

Signed-off-by: Rafa Porres Molina <rporresm@redhat.com>

### CODE REVIEW

**Comment by rporres** [comment]
A few notes:

* Missing documentation. I will create once we like the approach
* local `hack/generate-bundle-contents.py` script should be moved to boilerplate. It should follow the conventions from https://github.com/openshift/boilerplate/pull/101
* Previous `docker-push` was unauthenticated. I made it depend from login and build. This is a breaking change
* Tested in deployment-validation-operator

File: boilerplate/openshift/golang-osd-operator/standard.mk
Comment (2uasimojo) [inline boilerplate/openshift/golang-osd-operator/standard.mk:99]: Is there a way to check whether already authenticated? If so, it would be preferable to do that, and only require/use the cred vars if necessary.

**Review by 2uasimojo — COMMENTED** [review]
Having already reviewed this code in dvo itself, I'll just say I'm on board with the overall concepts.

We discussed this in TΩ this morning and want to

/assign @bdematte 

to dig into how the various pieces should be split up. We want this and #101 to evolve together, and use common chunks where approriate.

File: boilerplate/openshift/golang-osd-operator/standard.mk
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/standard.mk:99]: The presence of docker configuration doesn't guarantee that it has a working set of credentials, so I don't see how this could be done easily.

On second thoughts, docker pushes must be protected as much as possible. They are one of the most delicate parts of the system and we know the consequences of bad pushes to both the catalog or the operator image... I don't think it is a bad thing that creds are always requested.

File: boilerplate/openshift/golang-osd-operator/standard.mk
Comment (bdematte) [inline boilerplate/openshift/golang-osd-operator/standard.mk:36]: Shouldn't it be gitlab.cee.redhat.com/service/saas-operator-versions.git ?

**Review by bdematte — CHANGES_REQUESTED** [review]

File: boilerplate/openshift/golang-osd-operator/standard.mk
Comment (2uasimojo) [reply boilerplate/openshift/golang-osd-operator/standard.mk:36]: +1 https://github.com/app-sre/deployment-validation-operator/blob/master/Makefile#L1

Comment (2uasimojo) [reply boilerplate/openshift/golang-osd-operator/standard.mk:36]: Though thinking about it, why not just hardcode it in the script itself, since it's going to be the same for every consumer? That'll save us a parameter/envvar :)

Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/standard.mk:36]: Good catch. Yes, I will change it. I would leave as parametrizable, I like the possibility of overriding it, e.g in testing scenarios.

Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/standard.mk]: Fixed in ea1c9f7

**Comment by rporres** [comment]
What's missing here? I understand `hack/generate-bundle-contents.py` should be generalized but I don't know if we want to wait until #101 is completed to move on with this PR.

**Comment by bdematte** [comment]
> What's missing here? I understand `hack/generate-bundle-contents.py` should be generalized but I don't know if we want to wait until #101 is completed to move on with this PR.

For me, we are missing documentation and testing for the PR to be mergeable. 
I don't think we should include more changes in this one to start - we can go with the rest (spliting it, generalizing the generate-bundle, and so on) in dedicated PRs.

**Comment by rrati** [comment]
@bdematte We've done testing with DVO for these changes.  What are you looking for re: testing and docs?

**Comment by rporres** [comment]
The documentation is missing, as stated before I didn't want to write it before we agreed with the approach.

What kind of testing do you want to see here?

**Comment by rporres** [comment]
For example, the catalog created via `opm` has been thoroughly tested: https://issues.redhat.com/browse/APPSRE-2741

End to end tests for this would be very much desirable, but not easily achievable since they involve many moving parts.

**Comment by bdematte** [comment]
We need to have integrated CI testing for it as this will be evolving in the coming weeks/month. 
PR#101 contains a first version of the testing for the 'old way' and you can use part of it for the testing of this I believe.

**Comment by rporres** [comment]
A few questions/comments about tests to be able to come with a valid strategy:

1. Where would this tests run? In a VM or in a docker container. This is important because the script uses extensively docker (`opm` uses docker internally, we run the catalog to verify it returns the proper version) and that would mean running docker in docker, which is always a challenge to say the least
2. The code would need to be split in parts either via functions or makefile targets. Makefile targets would be in line with the tests outlined in #101 but since targets would need to be wrapped in another script (they share lots of variables) that would mean that a script that is called from a Makefile target would call another targets of the same Makefile instead of using the common dependency approach. That looks hacky...
3. Bundle generation can not be tested as it is outside the code
4. Which is the test runner? Can something as bats be used?

**Comment by bdematte** [comment]
CI is running those tests through docker indeed. 
Without splitting it, we should at least be able to test that we can create create CSV for staging/production, and that the SKIP mechanism reflect properly (in gitlab and in the CSV). There are other cases that would be good to add eventually, but I think this would be the bare minimum and wouldn't require a split. 
`driver` is pretty simple and is running any executable files detected in the `test/case`

**Comment by jharrington22** [comment]
@rporres I review this and it looks good. What @bdematte is asking for is some tests so we can ensure iterating on this moving forward is safe, but given that it would require running docker in docker its going to be difficult.

Looking at the script I think having the following would be enough and shouldn't be too much work
- validate that the `generate-bundle-contents.py` is generating what we expect each time
- validating the parsing of `bundle_versions_file`
- validating the catalog build (looks like thats already added in the script) - is a dry run enough?

I'm ok with the documentation coming after this PR since things are likely to change pretty quickly and there isn't adoption from SRE operators just yet, we'll likely have to write something help people migrate.

@bdematte @rporres I don't want to hold up this PR, I'd like it to merge before code freeze so I'm OK If the tests are very simple static validations to begin with.

**Comment by rporres** [comment]
@jharrington22 I've had a meeting with @bdematte to come with a strategy to have this merged before code freeze:

> validate that the generate-bundle-contents.py is generating what we expect each time

not feasible. That's a local script in `dvo` repo. I can move it into a target that has to be overriden in the boilerplate consumer to make it more flexible

> validating the parsing of bundle_versions_file

I will extract that logic into a separate script so that it can be tested

> validating the catalog build (looks like thats already added in the script) - is a dry run enough?

I don't see how to do this as validating the catalog involves running the container... Dry run mode has proven to be very helpful to me while developing, I think it should be enough

> @bdematte @rporres I don't want to hold up this PR, I'd like it to merge before code freeze so I'm OK If the tests are very simple static validations to begin with.

I think we have time before Wednesday 8pm to add some tests and review this. Let's try to do it :)

**Comment by jharrington22** [comment]
@rporres sounds great! Thank you.

**Comment by rporres** [comment]
In order to add tests I had to refactor the original code into pieces. Done in c78b5a7
Tests are in 47dfad7

File: test/case/convention/openshift/golang-osd-operator/05-build-olm-catalog
Comment (bdematte) [inline test/case/convention/openshift/golang-osd-operator/05-build-olm-catalog:17]: Shouldn't it be "gitlab.cee.redhat.com/service/saas-operator-versions.git" also here ?

Comment (rporres) [reply test/case/convention/openshift/golang-osd-operator/05-build-olm-catalog:17]: we don't really care... it is not used in the unit tests (and should not be used 😄 ). I will change it to be `not-important-at-the-moment` to further clarify it

Comment (rporres) [reply test/case/convention/openshift/golang-osd-operator/05-build-olm-catalog]: fixed in cc6b3f6

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (2uasimojo) [inline boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh:58]: 1. Use the param.
2. Do you need to make sure this command doesn't print anything to stdout, since you're relying on the output from the function?
```suggestion
    # Redirect to stderr to make sure this output doesn't interfere with stdout expected from the function
    "$REPO_ROOT"/boilerplate/openshift/golang-osd-operator/ensure.sh ${executable} >&2
```

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (2uasimojo) [inline boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh:37]: Since the bundle versions repo is so tightly bound to other variables that aren't "parameters" to this script (e.g. the path to the versions file), suggest simply hardcoding it and getting rid of this variable. (IMO the more we can reduce the number of env vars we're relying on, the better.)

**Review by 2uasimojo — COMMENTED** [review]
Partial review, more coming.

File: boilerplate/openshift/golang-osd-operator/standard.mk
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/standard.mk:101]: yes we want... let me take a look

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh:58]: 1. Yes. That's a bug. It was basically working because I had already everything installed
2. Yes. I wonder how that was working in the first place. I'll fix it

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh:37]: I was leaving this as a parameter because it's helpful to be able to test against a different git repository if needed. But I can remove it and make it overridable if needed to remove it from the Makefiles as it is just testing convenient.

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh:58]: 2. `ensure.sh` throws everything to stderr, that's why it was working. I'll add your suggestion as it is safer

Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh]: fixed in 3424552

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh]: Fixed in 1bd829f

**Review by 2uasimojo — COMMENTED** [review]
Partial review part deux.

File: boilerplate/openshift/golang-osd-operator/standard.mk
Comment (2uasimojo) [reply boilerplate/openshift/golang-osd-operator/standard.mk]: :+1:

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh]: fixed in 3e8ecf1

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh]: `OLM_BUNDLE_VERSIONS_REPO_BRANCH` fixed in 6a583f7

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh:162]: Yeah, that could be potentially dangerous... On the other hand, that's THE function that we have thoroughly tested 😂 

I'll take a look into that after we have merged this, to see if I can come with something in the lines you're suggesting

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh]: fixed in de4b43c

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh:19]: No need for that. It's only needed when you interpolate variables, not when you define them

File: boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh
Comment (rporres) [reply boilerplate/openshift/golang-osd-operator/build-opm-catalog.sh:158]: empty message has a different meaning... It means no previous version ever existed. I will take care of this after this is merged as I want to add a test case for it. I agree that `exit`s should only be in `main`

**Comment by rporres** [comment]
Squashed all the commits into one. I have left a couple of things to be done afterwards, as I'm running out of time today. We can totally live with them while I fix them.

**Review by 2uasimojo — APPROVED** [review]
I haven't finished reviewing, but James and Ben have been through it, and since it's not in the way of anything, we've agreed to land it now and iterate as needed.

/lgtm
/approve

**Pull request merged**
